tests/test_fipa_end2end.py

```diff
@@ -0,0 +1,307 @@
+            message: Message, receiver_address: Address
+        ) -> BaseDialogue.Role:
+            """Infer the role of the agent from an incoming/outgoing first message
+
+            :param message: an incoming/outgoing first message
+            :param receiver_address: the address of the receiving agent
+            :return: The role of the agent
+            """
+            return FipaDialogue.Role.BUYER
+
+        FipaDialogues.__init__(
+            self,
+            self_address=self_address,
+            role_from_first_message=role_from_first_message,
+            dialogue_class=FipaDialogue,
+        )
+
+
+class BuyerBehaviour(TickerBehaviour):
+    """Test buyer behaviour."""
+
+    addr: str
+
+    def setup(self) -> None:
+        """Set up behaviour."""
+        self.is_started = False
+        self.was_called = False
+
+    def act(self) -> None:
+        """Make an action."""
+        dialogues = cast(FipaDialogues, self.context.dialogues)
+        if not self.is_started or self.was_called:
+            return
+
+        cfp_msg, _ = dialogues.create(
+            counterparty=self.addr,
+            performative=FipaMessage.Performative.CFP,
+            query=Query([Constraint("something", ConstraintType(">", 1))]),
+        )
+        self.context.outbox.put_message(cfp_msg)
+        self.was_called = True
+
+    def start(self, addr: Address) -> None:
+        """Set counterpart addrtess and start sending CFP."""
+        self.addr = addr
+        self.is_started = True
+
+
+class BuyerHandler(Handler):
+    """Test BuyerHandler."""
+
+    SUPPORTED_PROTOCOL = FipaMessage.protocol_id
+
+    def setup(self) -> None:
+        """Set up behaviour."""
+        self.got_proposal = False
+
+    def teardown(self) -> None:
+        """Tear down handler."""
+
+    def handle(self, message) -> None:
+        """Handle an evelope."""
+        dialogues = cast(FipaDialogues, self.context.dialogues)
+        buyer_dialogue = dialogues.update(message)
+        if not buyer_dialogue:
+            return
+        self.got_proposal = True
+
+
+class SellerHandler(Handler):
+    """Simple behaviour to count how many acts were called."""
+
+    SUPPORTED_PROTOCOL = FipaMessage.protocol_id
+
+    def setup(self) -> None:
+        """Set up behaviour."""
+        self.dialogues: Optional[FipaDialogues] = None
+
+    def teardown(self) -> None:
+        """Tear down handler."""
+
+    def handle(self, message) -> None:
+        """Handle an evelope."""
+        if not self.dialogues:
+            self.dialogues = SellerDialogues(self.context.agent_address)
+
+        seller_dialogue = self.dialogues.update(message)
+        if not seller_dialogue:
+            return
+        proposal_msg = seller_dialogue.reply(
+            target_message=message,
+            performative=FipaMessage.Performative.PROPOSE,
+            proposal=Description({"foo1": 1, "bar1": 2}),
+        )
+        self.context.outbox.put_message(proposal_msg)
```

Comment:
Ideally we have the full fipa flow, so `cfp > propose > accept > match accept > inform`. But that can be done later and copy pasted from one of the skills.